### PR TITLE
SP-6321 opsgenie refactor

### PR DESCRIPTION
**Pull request by nshem**

### CODE REVIEW

**Comment by hagaishapira** [comment]
@nshem is this PR ready?

**Comment by hagaishapira** [comment]
@nshem Is this review ready ?

**Review by hagaishapira — CHANGES_REQUESTED** [review]
Please see that the manifests are missing outputs. Please address this and add relevant field to the output struct

File: steps/opsgenie/alert/acknowledge/acknowledge.go [hunk 1]
```diff
@@ -1,50 +1,54 @@
 package main
 
 import (
+	"context"
+
 	"github.com/Jeffail/gabs/v2"
-	envconf "github.com/caarlos0/env/v6"
 	"github.com/opsgenie/opsgenie-go-sdk-v2/alert"
```
Comment (oriser) [inline steps/opsgenie/alert/acknowledge/acknowledge.go:8]: delete empty line

File: steps/opsgenie/alert/create/main.go [hunk 2]
```diff
@@ -26,44 +29,49 @@ type OpsgenieCreateAlert struct {
...
-	return fmt.Errorf("invalid responder type, should be one of: 'user', 'team', 'escalation', 'schedule'")
-
+	return status.FromError(codes.InvalidArgument, "invalid responder type", fmt.Errorf("should be one of: 'user', 'team', 'escalation', 'schedule'"))
 }
 
-func (o *OpsgenieCreateAlert) Init() error {
+func (o *OpsgenieCreateAlert) Init() status.Status {
 	err := envconf.ParseWithFuncs(&o.args, map[reflect.Type]envconf.ParserFunc{
 		reflect.TypeOf(map[string]string{}): env.ParseKeyValueEnv,
 	})
```
Comment (oriser) [inline steps/opsgenie/alert/create/main.go:52]: That's not good, we should use only the SDK `env.Parse`, as it contains more than just calling this library.
Move that type of parsing func to the SDK

File: steps/opsgenie/incident/add-note/Dockerfile [hunk 3]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/incident/add-note/Dockerfile:11]: Change the entrypoint too

File: steps/opsgenie/alert/create/Dockerfile [hunk 4]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/alert/create/Dockerfile:11]: Change the entrypoint

File: steps/opsgenie/incident/add-note/add-note.go [hunk 5]
```diff
@@ -1,53 +1,57 @@
 package main
 
 import (
+	"context"
+
 	"github.com/Jeffail/gabs/v2"
-	envconf "github.com/caarlos0/env/v6"
 	"github.com/opsgenie/opsgenie-go-sdk-v2/incident"
```
Comment (oriser) [inline steps/opsgenie/incident/add-note/add-note.go:8]: empty line

File: steps/opsgenie/incident/close/Dockerfile [hunk 6]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/incident/close/Dockerfile:11]: entrypoint

File: steps/opsgenie/incident/close/close.go [hunk 7]
```diff
@@ -1,54 +1,57 @@
 package main
 
 import (
+	"context"
+
 	"github.com/Jeffail/gabs/v2"
-	envconf "github.com/caarlos0/env/v6"
 	"github.com/opsgenie/opsgenie-go-sdk-v2/incident"
```
Comment (oriser) [inline steps/opsgenie/incident/close/close.go:8]: empty line

File: steps/opsgenie/incident/create/Dockerfile [hunk 8]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/incident/create/Dockerfile:11]: entrypoint

File: steps/opsgenie/incident/resolve/Dockerfile [hunk 9]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/incident/resolve/Dockerfile:11]: entrypoint

File: steps/opsgenie/incident/create/main.go [hunk 10]
```diff
@@ -1,12 +1,14 @@
 package main
 
 import (
-	"fmt"
+	"context"
```
Comment (oriser) [inline steps/opsgenie/incident/create/main.go:5]: empty line

File: steps/opsgenie/incident/resolve/resolve.go [hunk 11]
```diff
@@ -1,81 +1,85 @@
...
 import (
+	"context"
+	"errors"
 	"net/http"
 
 	"github.com/Jeffail/gabs/v2"
-	envconf "github.com/caarlos0/env/v6"
 	"github.com/opsgenie/opsgenie-go-sdk-v2/client"
 	"github.com/opsgenie/opsgenie-go-sdk-v2/incident"
-	"github.com/pkg/errors"
```
Comment (oriser) [inline steps/opsgenie/incident/resolve/resolve.go:11]: empty lne

File: steps/opsgenie/oncall/get/Dockerfile [hunk 12]
```diff
@@ -8,7 +8,7 @@ ARG STEP_BASEPATH
 FROM us-docker.pkg.dev/stackpulse/public/base:$BASE_BRANCH AS base
 
 # 2. Builder
-FROM golang:1.14-alpine as builder
+FROM golang:1.16-alpine as builder
```
Comment (oriser) [inline steps/opsgenie/oncall/get/Dockerfile:11]: entrypoint

**Review by oriser — CHANGES_REQUESTED** [review]

File: steps/opsgenie/oncall/get/main.go [hunk 13]
```diff
@@ -18,39 +21,37 @@ type OpsgenieOncallList struct {
 	args Args
 }
 
-func (o *OpsgenieOncallList) Init() error {
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieOncallList) Init() status.Status {
+	stat := env.Parse(&o.args)
```
Comment (hagaishapira) [inline steps/opsgenie/oncall/get/main.go:25]: can return env.Parse directly

File: steps/opsgenie/incident/resolve/resolve.go [hunk 14]
```diff
@@ -1,81 +1,85 @@
...
 type OpsgenieIncidentResolve struct {
 	args Args
 }
 
-func (o *OpsgenieIncidentResolve) Init() error {
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieIncidentResolve) Init() step.Status {
+	stat := env.Parse(&o.args)
```
Comment (hagaishapira) [inline steps/opsgenie/incident/resolve/resolve.go:30]: return env.Parse directly

File: steps/opsgenie/incident/create/main.go [hunk 15]
```diff
@@ -23,39 +25,41 @@ type OpsgenieIncidentCreate struct {
...
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieIncidentCreate) Init() status.Status {
+	stat := env.Parse(&o.args)
+	if stat.Failed() {
+		return stat
+	}
+
+	if err := incident.ValidatePriority(incident.Priority(o.args.Priority)); err != nil {
```
Comment (hagaishapira) [inline steps/opsgenie/incident/create/main.go:34]: do you need this with the new oneof?

Comment (hagaishapira) [reply steps/opsgenie/incident/create/main.go:34]: Check out the change I had here for the oneof in the PR I had, use what you've created in the SDK :)

File: steps/opsgenie/incident/close/close.go [hunk 16]
```diff
@@ -1,54 +1,56 @@
...
 type OpsgenieIncidentClose struct {
 	args Args
 }
 
-func (o *OpsgenieIncidentClose) Init() error {
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieIncidentClose) Init() status.Status {
+	stat := env.Parse(&o.args)
```
Comment (hagaishapira) [inline steps/opsgenie/incident/close/close.go:26]: return Parse directly.

File: steps/opsgenie/incident/add-note/add-note.go [hunk 17]
```diff
@@ -1,53 +1,56 @@
...
 type OpsgenieIncidentAddNote struct {
 	args Args
 }
 
-func (o *OpsgenieIncidentAddNote) Init() error {
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieIncidentAddNote) Init() status.Status {
+	stat := env.Parse(&o.args)
```
Comment (hagaishapira) [inline steps/opsgenie/incident/add-note/add-note.go:26]: return Parse directly

File: steps/opsgenie/alert/create/main.go [hunk 18]
```diff
@@ -26,44 +27,48 @@ type OpsgenieCreateAlert struct {
...
 
-	if err != nil {
-		return err
+	stat = validateResponder(o.args)
+	if stat.Failed() {
+		return stat
 	}
 
-	err = validateResponder(o.args)
+	err := alert.ValidatePriority(alert.Priority(o.args.Priority))
```
Comment (hagaishapira) [inline steps/opsgenie/alert/create/main.go:58]: can use oneof and change the manifest with the SDK validator.

File: steps/opsgenie/alert/acknowledge/acknowledge.go [hunk 19]
```diff
@@ -1,50 +1,53 @@
...
 type OpsgenieAcknowledgeAlert struct {
 	args Args
 }
 
-func (o *OpsgenieAcknowledgeAlert) Init() error {
-	err := envconf.Parse(&o.args)
-	if err != nil {
-		return err
+func (o *OpsgenieAcknowledgeAlert) Init() status.Status {
+	stat := env.Parse(&o.args)
```
Comment (hagaishapira) [inline steps/opsgenie/alert/acknowledge/acknowledge.go:25]: return Parse directly

**Review by hagaishapira — CHANGES_REQUESTED** [review]
See comments

File: steps/opsgenie/alert/create/main.go (hunk 18 above)
Comment (nshem) [reply steps/opsgenie/alert/create/main.go:58]: one of is static. this is a dynamic validation.

**Review by hagaishapira — APPROVED** [review]
Update manifests and that's it

**Review by oriser — APPROVED** [review]

**Pull request merged**